crates/modelardb_compression/src/lib.rs

```diff
@@ -32,62 +32,155 @@ use modelardb_common::types::{
-    // reallocation as one model is created per data point in the worst case.
+    // Enough memory for end_index compressed segments are allocated to never require reallocation
+    // as one compressed segments is created per data point in the absolute worst case.
     let end_index = uncompressed_timestamps.len();
     let mut compressed_record_batch_builder = CompressedSegmentBatchBuilder::new(end_index);
 
     // Compress the uncompressed timestamps and uncompressed values.
     let mut current_index = 0;
+    let mut residual_start_index = 0;
     while current_index < end_index {
-        // Create a compressed segment that represents the data points from
-        // current_index to index within error_bound where index <= end_index.
-        let compressed_segment_builder = CompressedSegmentBuilder::new(
-            univariate_id,
+        // Propose a model that represents the values from current_index to index within error_bound
+        // where index <= end_index.
+        let compressed_model_builder = CompressedModelBuilder::new(
             current_index,
             end_index,
             uncompressed_timestamps,
             uncompressed_values,
             error_bound,
         );
-        current_index = compressed_segment_builder.finish(&mut compressed_record_batch_builder);
+
+        let selected_model = compressed_model_builder.finish();
+
+        // The selected model is only stored as part of a compressed segment if it uses less storage
+        // space per value than the uncompressed values it represents.
+        if selected_model.bytes_per_value <= models::VALUE_SIZE_IN_BYTES as f32 {
+            // Compress residual data points with gorilla and store them in a compressed segments.
+            if residual_start_index != current_index {
+                compress_and_store_residual_value_range(
+                    univariate_id,
+                    residual_start_index,
+                    current_index - 1,
+                    uncompressed_timestamps,
+                    uncompressed_values,
+                    &mut compressed_record_batch_builder,
+                )
+            }
+
+            // Store the selected model and the corresponding timestamps in a compressed segments.
+            store_selected_model_in_batch_builder(
+                univariate_id,
+                current_index,
+                &selected_model,
+                uncompressed_timestamps,
+                &mut compressed_record_batch_builder,
+            );
+
+            // Update the indexes to specify that all data points until now has been compressed.
+            current_index = selected_model.end_index + 1;
+            residual_start_index = current_index;
+        } else {
+            // The potentially lossy models could not efficiently encode the sub-sequence starting
+            // at current_index, the residual values will instead be compressed using gorilla.
```

Comment:
Since we are referring to the model `Gorilla`, maybe it should be capitalized. This might also need to be changed on line 84 and on line 115.

Reply:
Fixed.